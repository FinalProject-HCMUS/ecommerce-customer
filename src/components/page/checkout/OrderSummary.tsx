import type React from 'react';
import { ProductItem } from './ProductItem';
import Divider from '../../shared/Divider';
import { GeneralButton } from '../../shared/Button';

export const OrderSummary: React.FC = () => {
  const products = [
    {
      id: 1,
      name: 'Checkered Shirt',
      size: 'Medium',
      color: 'Red',
      price: 180,
      quantity: 1,
      image:
        'https://res.cloudinary.com/dt0ps34k9/image/upload/v1743842005/shirt_ckwim3.png',
    },
    {
      id: 2,
      name: 'Checkered Shirt',
      size: 'Medium',
      color: 'Red',
      price: 180,
      quantity: 1,
      image:
        'https://res.cloudinary.com/dt0ps34k9/image/upload/v1743842005/shirt_ckwim3.png',
    },
    {
      id: 3,
      name: 'Checkered Shirt',
      size: 'Medium',
      color: 'Red',
      price: 180,
      quantity: 1,
      image:
        'https://res.cloudinary.com/dt0ps34k9/image/upload/v1743842005/shirt_ckwim3.png',
    },
  ];

  const subtotal = products.reduce(
    (sum, product) => sum + product.price * product.quantity,
    0
  );
  const discountRate = 0.2; // 20%
  const discountAmount = Math.round(subtotal * discountRate);
  const deliveryFee = 15;
  const total = subtotal - discountAmount + deliveryFee;

  return (
    <div>
      <div className="space-y-4 mb-8">
        {products.map((product) => (
          <ProductItem key={product.id} product={product} />
        ))}
      </div>

      <Divider />

      <div className="pt-6">
        <h2 className="text-xl font-bold mb-4">Order Summary</h2>

        <div className="space-y-3">
          <div className="flex justify-between">
            <span className="text-gray-600">Subtotal</span>
            <span className="font-medium">${subtotal}</span>
          </div>

          <div className="flex justify-between">
            <span className="text-gray-600">Discount (-20%)</span>
            <span className="font-medium text-red-500">-${discountAmount}</span>
          </div>

          <div className="flex justify-between">
            <span className="text-gray-600">Delivery Fee</span>
            <span className="font-medium">${deliveryFee}</span>
          </div>

          <Divider />

          <div className="flex justify-between pt-3">
            <span className="font-bold">Total</span>
            <span className="font-bold text-xl">${total}</span>
          </div>
        </div>

        <GeneralButton
          type="submit"
          className="w-full mt-6 bg-black text-white py-4 px-6 rounded-full font-medium hover:bg-gray-800 transition-colors"
        >
          MAKE PAYMENT
        </GeneralButton>
      </div>
    </div>
  );
};
