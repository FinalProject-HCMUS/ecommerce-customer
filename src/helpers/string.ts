const formatCurrency = (value: number, currency: 'USD' | 'VND'): string => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    maximumFractionDigits: 0,
  }).format(value);
};

export { formatCurrency };
