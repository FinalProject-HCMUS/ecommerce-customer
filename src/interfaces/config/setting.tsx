export interface SystemSettingResponse {
  id: string;
  key: string;
  value: object;
  serviceName: string;
}
