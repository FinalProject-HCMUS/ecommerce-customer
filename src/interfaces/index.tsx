export * from './blog/BlogResponse'
export * from './category/CategoryResponse'
export * from './message/MessageResponse'
export * from './order/OrderDetailResponse'
export * from './order/OrderResponse'
export * from './order/OrderTrackResponse'
export * from './order/PaymentMethod'
export * from './order/Status'
export * from './product/CartItemResponse'
export * from './product/ColorResponse'
export * from './product/ProductColorSizeResponse'
export * from './product/ProductImageResponse'
export * from './product/ProductResponse'
export * from './product/TopProduct'
export * from './product/SizeResponse'
export * from './review/ReviewResponse'
export * from './user/UserResponse'
export * from './user/CreateUserRequest'
export * from './user/Role'
export * from './region/region'
export * from './common/CustomResponse'
export * from './common/Pageable'
export * from './common/CustomError'
