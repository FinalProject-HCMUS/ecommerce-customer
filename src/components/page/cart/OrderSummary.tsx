import type React from 'react';
import { ArrowRight } from 'lucide-react';
import type { OrderSummaryProps } from '../../../interfaces/temp/cart';
import { t } from '../../../helpers/i18n';
import { formatCurrency } from '../../../helpers/string';

const OrderSummary: React.FC<OrderSummaryProps> = ({ data }) => {
  const { subtotal, deliveryFee, total } = data;

  return (
    <div className="bg-white rounded-[12px] p-6 border border-gray-200">
      <h2 className="text-xl font-bold mb-6">{t('lbl.orderSummary')}</h2>

      <div className="space-y-4">
        <div className="flex justify-between">
          <span className="text-gray-600">{t('lbl.subTotal')}</span>
          <span className="font-medium">{formatCurrency(subtotal, 'VND')}</span>
        </div>

        <div className="flex justify-between">
          <span className="text-gray-600">{t('lbl.deliveryFee')}</span>
          <span className="font-medium">
            {formatCurrency(deliveryFee, 'VND')}
          </span>
        </div>

        <div className="border-t border-gray-200 pt-4 mt-4">
          <div className="flex justify-between">
            <span className="text-lg font-medium">{t('lbl.total')}</span>
            <span className="text-lg font-bold">
              {formatCurrency(total, 'VND')}
            </span>
          </div>
        </div>

        <button className="w-full bg-black text-white py-4 px-6 rounded-full mt-6 flex items-center justify-center">
          <span>{t('lbl.gotoCheckout')}</span>
          <ArrowRight className="ml-2" size={18} />
        </button>
      </div>
    </div>
  );
};

export default OrderSummary;
