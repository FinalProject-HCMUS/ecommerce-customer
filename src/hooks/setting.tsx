import { useState, useCallback } from 'react';
import SettingsService from '../services/apis/setting';
import { SystemSettingResponse } from '../interfaces/config/setting';

export const useSystemSettings = (serviceName?: string) => {
  const [settings, setSettings] = useState<SystemSettingResponse[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const fetchSettings = useCallback(
    async (serviceNameFilter?: string) => {
      setLoading(true);
      try {
        const response = await SettingsService.getAllSystemSettings(
          serviceNameFilter || serviceName
        );
        setSettings(response);
      } finally {
        setLoading(false);
      }
    },
    [serviceName]
  );

  return {
    settings,
    loading,
    fetchSettings,
  };
};

export default useSystemSettings;
