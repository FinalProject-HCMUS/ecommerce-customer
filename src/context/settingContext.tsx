import React, {
  createContext,
  useContext,
  useEffect,
  useState,
  ReactNode,
} from 'react';
import SettingsService from '../services/apis/setting';
import { SystemSettingResponse } from '../interfaces/config/setting';

interface SettingsContextType {
  settings: SystemSettingResponse[];
  loading: boolean;
  error: string | null;
  refreshSettings: () => Promise<void>;
}

const SettingsContext = createContext<SettingsContextType | undefined>(
  undefined
);

export const useSettingsContext = () => {
  const context = useContext(SettingsContext);
  if (context === undefined) {
    throw new Error(
      'useSettingsContext must be used within a SettingsProvider'
    );
  }
  return context;
};

interface SettingsProviderProps {
  children: ReactNode;
}

export const SettingsProvider: React.FC<SettingsProviderProps> = ({
  children,
}) => {
  const [settings, setSettings] = useState<SystemSettingResponse[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);

  const fetchSettings = async () => {
    setLoading(true);
    setError(null);
    try {
      const response = await SettingsService.getAllSystemSettings('MySetting');
      setSettings(response);
    } catch (err) {
      console.error('Failed to fetch system settings:', err);
      setError('Failed to load system settings');
    } finally {
      setLoading(false);
    }
  };

  // Initial fetch on mount
  useEffect(() => {
    fetchSettings();
  }, []);

  const value = {
    settings,
    loading,
    error,
    refreshSettings: fetchSettings,
  };

  return (
    <SettingsContext.Provider value={value}>
      {children}
    </SettingsContext.Provider>
  );
};
