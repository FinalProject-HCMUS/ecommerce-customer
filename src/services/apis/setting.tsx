import client from './request';
import { SystemSettingResponse } from '../../interfaces/config/setting';

export const SettingsService = {
  getAllSystemSettings: async (
    serviceName?: string
  ): Promise<SystemSettingResponse[]> => {
    const params = serviceName ? { serviceName } : {};
    const response = await client.get('/system-settings', { params });
    return response.data;
  },

  getAllServiceNames: async (): Promise<string[]> => {
    const response = await client.get('/system-settings/service-names');
    return response.data;
  },
};

export default SettingsService;
