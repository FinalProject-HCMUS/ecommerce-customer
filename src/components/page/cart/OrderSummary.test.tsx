import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import OrderSummary from './OrderSummary';
import { OrderSummaryData } from '../../../interfaces';
import * as settingContext from '../../../context/settingContext';
import * as stringHelper from '../../../helpers/string';
import { SystemSettingResponse } from '../../../interfaces/config/setting';

// Mock dependencies
jest.mock('../../../helpers/i18n', () => ({
  t: jest.fn((key) => key),
}));

jest.mock('../../../helpers/string', () => ({
  formatCurrency: jest.fn((value, currency) => 
    currency === 'USD' ? `$${value}` : `${value} â‚«`
  ),
}));

jest.mock('../../../context/settingContext', () => ({
  useSettingsContext: jest.fn(),
}));

describe('OrderSummary', () => {
  const mockOrderData: OrderSummaryData = {
    subtotal: 200000,
    deliveryFee: 20000,
    total: 220000,
  };


    const mockSettings : SystemSettingResponse[]= [
    { 
      id: '1', 
      key: 'CurrencyCode', 
      value: 'VND' as unknown as object, 
      serviceName: 'MySetting'  // Changed from 'group' to 'serviceName'
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Setup default mocks
    jest.spyOn(settingContext, 'useSettingsContext').mockReturnValue({
      settings: mockSettings,
      loading: false,
      error: null,
      refreshSettings: jest.fn(),
    });
  });

  test('renders with required props and VND currency', () => {
    render(<OrderSummary data={mockOrderData} />);
    
    // Check headings and labels
    expect(screen.getByText('lbl.orderSummary')).toBeInTheDocument();
    expect(screen.getByText('lbl.subTotal')).toBeInTheDocument();
    expect(screen.getByText('lbl.deliveryFee')).toBeInTheDocument();
    expect(screen.getByText('lbl.total')).toBeInTheDocument();
    
    // Check formatted currency values
    expect(stringHelper.formatCurrency).toHaveBeenCalledWith(200000, 'VND');
    expect(stringHelper.formatCurrency).toHaveBeenCalledWith(20000, 'VND');
    expect(stringHelper.formatCurrency).toHaveBeenCalledWith(220000, 'VND');
    
    // Check button is enabled by default
    const checkoutButton = screen.getByText('lbl.gotoCheckout').closest('button');
    expect(checkoutButton).not.toBeDisabled();
  });

  test('does not render item counts when not provided', () => {
    render(<OrderSummary data={mockOrderData} />);
    
    // Should not find the selected items text
    expect(screen.queryByText('lbl.selectedItems:')).not.toBeInTheDocument();
  });

  test('handles checkout button click', () => {
    const handleCheckout = jest.fn();
    render(<OrderSummary data={mockOrderData} onCheckout={handleCheckout} />);
    
    // Click the checkout button
    const checkoutButton = screen.getByText('lbl.gotoCheckout').closest('button');
    fireEvent.click(checkoutButton!);
    
    // Check if the callback was called
    expect(handleCheckout).toHaveBeenCalledTimes(1);
  });

  test('disables checkout button when isCheckoutEnabled is false', () => {
    render(<OrderSummary data={mockOrderData} isCheckoutEnabled={false} />);
    
    // Check if the button is disabled
    const checkoutButton = screen.getByText('lbl.gotoCheckout').closest('button');
    expect(checkoutButton).toBeDisabled();
    expect(checkoutButton).toHaveStyle('opacity: 0.5');
    expect(checkoutButton).toHaveStyle('cursor: not-allowed');
  });

  test('uses default VND currency when currency setting is not found', () => {
    jest.spyOn(settingContext, 'useSettingsContext').mockReturnValue({
      settings: [], // Empty settings, no currency code
      loading: false,
      error: null,
      refreshSettings: jest.fn(),
    });
    
    render(<OrderSummary data={mockOrderData} />);
    
    // Should default to VND
    expect(stringHelper.formatCurrency).toHaveBeenCalledWith(200000, 'VND');
  });

  test('matches snapshot', () => {
    const { container } = render(<OrderSummary data={mockOrderData} />);
    expect(container).toMatchSnapshot();
  });
});