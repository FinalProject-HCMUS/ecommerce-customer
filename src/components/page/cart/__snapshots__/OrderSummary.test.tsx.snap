// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderSummary matches snapshot 1`] = `
<div>
  <div
    class="bg-white rounded-[12px] p-6 border border-gray-200"
  >
    <h2
      class="text-xl font-bold mb-6"
    >
      lbl.orderSummary
    </h2>
    <div
      class="space-y-4"
    >
      <div
        class="flex justify-between"
      >
        <span
          class="text-gray-600"
        >
          lbl.subTotal
        </span>
        <span
          class="font-medium"
        >
          200000 ₫
        </span>
      </div>
      <div
        class="flex justify-between"
      >
        <span
          class="text-gray-600"
        >
          lbl.deliveryFee
        </span>
        <span
          class="font-medium"
        >
          20000 ₫
        </span>
      </div>
      <div
        class="border-t border-gray-200 pt-4 mt-4"
      >
        <div
          class="flex justify-between"
        >
          <span
            class="text-lg font-medium"
          >
            lbl.total
          </span>
          <span
            class="text-lg font-bold"
          >
            220000 ₫
          </span>
        </div>
      </div>
      <button
        class="w-full bg-black text-white py-4 px-6 rounded-full mt-6 flex items-center justify-center"
      >
        <span>
          lbl.gotoCheckout
        </span>
        <svg
          class="lucide lucide-arrow-right ml-2"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="m12 5 7 7-7 7"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
